crates/utils/src/tracing/grpc.rs

```diff
@@ -26,25 +33,68 @@ pub fn block_producer_trace_fn<T>(request: &http::Request<T>) -> tracing::Span {
-        _ => tracing::info_span!("store.rpc/Unknown"),
+        Some("ApplyBlock") => rpc_span!("store.rpc", "ApplyBlock"),
+        Some("CheckNullifiers") => rpc_span!("store.rpc", "CheckNullifiers"),
+        Some("CheckNullifiersByPrefix") => rpc_span!("store.rpc", "CheckNullifiersByPrefix"),
+        Some("GetAccountDetails") => rpc_span!("store.rpc", "GetAccountDetails"),
+        Some("GetAccountProofs") => rpc_span!("store.rpc", "GetAccountProofs"),
+        Some("GetAccountStateDelta") => rpc_span!("store.rpc", "GetAccountStateDelta"),
+        Some("GetBlockByNumber") => rpc_span!("store.rpc", "GetBlockByNumber"),
+        Some("GetBlockHeaderByNumber") => rpc_span!("store.rpc", "GetBlockHeaderByNumber"),
+        Some("GetBlockInputs") => rpc_span!("store.rpc", "GetBlockInputs"),
+        Some("GetBatchInputs") => rpc_span!("store.rpc", "GetBatchInputs"),
+        Some("GetNotesById") => rpc_span!("store.rpc", "GetNotesById"),
+        Some("GetTransactionInputs") => rpc_span!("store.rpc", "GetTransactionInputs"),
+        Some("SyncNotes") => rpc_span!("store.rpc", "SyncNotes"),
+        Some("SyncState") => rpc_span!("store.rpc", "SyncState"),
+        _ => rpc_span!("store.rpc", "Unknown"),
     };
 
-    span.set_parent(request);
-    span.set_http_attributes(request);
+    add_otel_span_attributes(span, request)
+}
+
+/// Adds remote tracing context to the span.
+///
+/// Could be expanded in the future by adding in more open-telemetry properties.
+fn add_otel_span_attributes<T>(span: tracing::Span, request: &http::Request<T>) -> tracing::Span {
+    use super::OpenTelemetrySpanExt;
+    // Pull the open-telemetry parent context using the HTTP extractor. We could make a more
+    // generic gRPC extractor by utilising the gRPC metadata. However that
+    //     (a) requires cloning headers,
+    //     (b) we would have to write this ourselves, and
+    //     (c) gRPC metadata is transferred using HTTP headers in any case.
+    let otel_ctx = opentelemetry::global::get_text_map_propagator(|propagator| {
+        propagator.extract(&MetadataExtractor(&tonic::metadata::MetadataMap::from_headers(
+            request.headers().clone(),
+        )))
+    });
+    tracing_opentelemetry::OpenTelemetrySpanExt::set_parent(&span, otel_ctx);
+
+    // Set HTTP attributes.
+    // See https://opentelemetry.io/docs/specs/semconv/rpc/rpc-spans/#server-attributes.
+    span.set_attribute("rpc.system", "grpc");
+    if let Some(host) = request.uri().host() {
+        span.set_attribute("server.address", host);
+    }
+    if let Some(host_port) = request.uri().port() {
+        span.set_attribute("server.port", host_port.as_str());
+    }
+    let remote_addr = request
+        .extensions()
+        .get::<tonic::transport::server::TcpConnectInfo>()
+        .and_then(tonic::transport::server::TcpConnectInfo::remote_addr);
+    if let Some(addr) = remote_addr {
+        span.set_attribute("client.address", addr.ip());
```

Comment:
This is the one that the spec says _may_ be the DNS name if no DNS lookup is required.

> Client address - domain name if available without reverse DNS lookup; otherwise, IP address or Unix domain socket name.

I'm wondering if we shouldn't use the `text_map_propagator` to cheat the system a bit, and on the client side inject `block-producer` and `rpc` as an additional property.

We could then use that here, and otherwise default to the IP if its missing.

Could also be done as a follow-up PR - I imagine you might be very much over this back-and-forth :D

Reply:
> Could also be done as a follow-up PR - I imagine you might be very much over this back-and-forth :D

Not at all - happy to iterate on this PR as much as required.

>I'm wondering if we shouldn't use the text_map_propagator to cheat the system a bit, and on the client side inject block-producer and rpc as an additional property.
We could then use that here, and otherwise default to the IP if its missing.

I'm a bit dubious as to the utility of that functionality but happy to add that to this PR if you think it is worthwhile.

Reply:
The main reason I'm considering it is because our components can (and currently do) run on the same server instance (and same process). Which means for the inter-component comms, the client and server IP address will just be the same since they're the same for all three components.

So I think it would help having the "caller" identifiable at least. Though maybe instead we should be hosting different api endpoints for this internal communication in any case.

I'm happy to just punt this to a separate issue instead.

Reply:
Ah ok yea that makes sense. Thanks for elaborating